### PR TITLE
ConnectivityStatus

### PR DESCRIPTION
**Pull request by 257**
this will help getting more granular reporting of connectivity status
from kernel.  it is linux/android specific for now with possible
future support for bsd/darwin.

this is very much a prototype; cleaner interface and code will follow.

it adds libnl (ships with android) dependency which is really not
required, although expedited prototyping.

### CODE REVIEW

File: cmake/FindNL.cmake [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+	set(HAVE_LIBNL 1)
+else()
+	set(NL_LIBRARIES )
+	set(NL_INCLUDE_DIRS)
+endif()
+
+MARK_AS_ADVANCED( NL_LIBRARIES NL_INCLUDE_DIRS )
+
+# MESSAGE(STATUS "NL_LIBRARIES ${NL_LIBRARIES}")
+# MESSAGE(STATUS "NL_INCLUDE_DIRS ${NL_INCLUDE_DIRS}")
```
Comment (AmarOk1412) [inline cmake/FindNL.cmake]: Debug message? Not necessary anymore

File: include/opendht/connstat.h [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+struct nl_sock;
+struct nl_msg;
+
+namespace dht {
+namespace net {
+
+class OPENDHT_PUBLIC ConnectivityStatus
+{
+public:
+    enum topic {
```
Comment (AmarOk1412) [inline include/opendht/connstat.h]: enum class Topic

File: include/opendht/connstat.h [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+    using connstat_event_cb = std::function<void (unsigned int)>;
+
+    void setTopicListener    (connstat_event_cb ucb, unsigned int topic);
+    void removeTopicListener (unsigned int topic);
+
+    int  nl_event_cb         (struct nl_msg*);
+
+private:
+    enum status { SUCCESS, FAILURE };
```
Comment (AmarOk1412) [inline include/opendht/connstat.h]: bool

File: src/connstat.cpp [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+
+/* forward declarations */
+static inline uint32_t nl_mgrp(uint32_t group);
+static int nlcb(struct nl_msg*, void*);
+
+ConnectivityStatus::ConnectivityStatus()
+    : nlsk(nlsk_init())
+    , logger_(new dht::Logger)
+{
+    if (nlsk_setup(nlsk.get()) == FAILURE)
```
Comment (AmarOk1412) [inline src/connstat.cpp]: if (not nlsk_setup(nlsk.get()))

File: include/opendht/connstat.h [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+
+    connstat_event_cb topic_cbs[CONNSTAT_TOPICS_MAX] = {nullptr};
```
Comment (AmarOk1412) [inline include/opendht/connstat.h]: std::array is here

Comment (AmarOk1412) [reply include/opendht/connstat.h]: {nullptr} => {}

File: src/connstat.cpp [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+    topic_cbs[topic] = ucb;
+}
+
+void
+ConnectivityStatus::removeTopicListener(unsigned int topic)
+{
+    switch (topic) {
+    case NEWADDR:
+        nl_socket_drop_memberships(nlsk.get(), RTNLGRP_IPV6_IFADDR, RTNLGRP_NONE);
+        topic_cbs[topic] = nullptr;
```
Comment (AmarOk1412) [inline src/connstat.cpp]: nullptr => {}

File: src/connstat.cpp [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+}
+
+void
+ConnectivityStatus::executer(unsigned int topic)
+{
+    if (*(topic_cbs+topic))
+        (*(topic_cbs+topic))(topic);
+}
+
+int
```
Comment (AmarOk1412) [inline src/connstat.cpp]: bool

Comment (AmarOk1412) [reply src/connstat.cpp]: unless if you will return a enum status, (if so just return a status)

File: src/connstat.cpp [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+     * Notifications do not use sequence numbers, disable sequence number
+     * checking.
+     */
+    nl_socket_disable_seq_check(nlsk);
+
+    /*
+     * Define a callback function, which will be called for each notification
+     * received
+     */
+    nl_socket_modify_cb(nlsk, NL_CB_VALID, NL_CB_CUSTOM, nlcb, (void *)this);
```
Comment (AmarOk1412) [inline src/connstat.cpp]: Just for info: https://en.cppreference.com/w/cpp/language/explicit_cast

Comment (AmarOk1412) [reply src/connstat.cpp]: Same for other cast

File: src/connstat.cpp [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+    return nl_mgrp(topic);
+}
+
+
+
+static inline uint32_t nl_mgrp(uint32_t group)
+{
+    if (group > 31 ) {
+        fprintf(stderr, "Use setsockopt for this group %d\n", group);
+        exit(-1);
```
Comment (AmarOk1412) [inline src/connstat.cpp]: cpp code, no C :) (replace exit, fprintf)

File: cmake/FindNL.cmake (hunk 1 above)
Comment (257) [reply cmake/FindNL.cmake]: done

File: src/connstat.cpp (hunk 9 above)
Comment (257) [reply src/connstat.cpp]: removed

File: src/connstat.cpp (hunk 8 above)
Comment (257) [reply src/connstat.cpp]: good read, thanks.

File: src/connstat.cpp (hunk 4 above)
Comment (257) [reply src/connstat.cpp]: done

File: include/opendht/connstat.h (hunk 2 above)
Comment (257) [reply include/opendht/connstat.h]: done

File: configure.ac [hunk 10]
```diff
@@ -18,8 +18,9 @@ AC_SUBST(OPENDHT_PATCH_VERSION, opendht_patch_version)
 
 AC_ARG_ENABLE([debug], AS_HELP_STRING([--enable-debug], [Build in debug mode, adds stricter warnings, disables optimization]))
 AS_IF([test "x$enable_debug" = "xyes"],
-      [CXXFLAGS="${CXXFLAGS} -g -Wno-return-type -Wall -Wextra -Wnon-virtual-dtor -O0 -pedantic-errors"],
-      [CXXFLAGS="${CXXFLAGS} -O3 -pedantic-errors"])
+      [
+       CXXFLAGS="${CXXFLAGS} -g -Wno-return-type -Wall -Wextra -Wnon-virtual-dtor -O0 -pedantic-errors -pedantic -ggdb -fsanitize=address"],
+       [CXXFLAGS="${CXXFLAGS} -O3 -pedantic-errors"])
```
Comment (aberaud) [inline configure.ac]: cleanup debug stuff

File: include/opendht/connstat.h [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+struct nl_sock;
+struct nl_msg;
+
+namespace dht {
+namespace net {
+
+class OPENDHT_PUBLIC ConnectivityStatus
+{
+public:
+    enum class topic : const long unsigned int {
```
Comment (aberaud) [inline include/opendht/connstat.h]: `Topic` (Use CapitalizedCamelCase for type names)
`const long unsigned int` should not be required (if it is required, just use `unsigned long`)

File: include/opendht/connstat.h [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+        TOPICS_DEL,
+        DELLINK,
+        DELROUTE4,
+        DELROUTE6,
+        DELROUTE,
+        DELADDR4,
+        DELADDR6,
+        DELADDR,
+        DELNEIGH,
+        TOPICS_MAX
```
Comment (aberaud) [inline include/opendht/connstat.h]: TOPICS_MAX should not be needed

File: include/opendht/connstat.h [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+    };
+
+    ConnectivityStatus();
+
+    using connstat_event_cb = std::function<void (long unsigned int)>;
```
Comment (aberaud) [inline include/opendht/connstat.h]: Use CapitalizedCamelCase for type names. Use Topic for the callback type: 

`using ConnectionEventCb = std::function<void(Topic)>;
`

File: include/opendht/connstat.h [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+
+    void setTopicListener    (connstat_event_cb ucb, topic);
+    void removeTopicListener (topic);
+
+    bool nl_event_cb         (struct nl_msg*);
+
+private:
+    bool SUCCESS, FAILURE;
```
Comment (aberaud) [inline include/opendht/connstat.h]: ???

File: include/opendht/connstat.h [hunk 15]
```diff
@@ -0,0 +1,97 @@
...
+
+    static const std::size_t topics_max = static_cast<std::size_t>(topic::TOPICS_MAX);
+    std::array<connstat_event_cb, topics_max> topic_cbs = {};
```
Comment (aberaud) [inline include/opendht/connstat.h]: use `std::map<Topic, ConnectivityEventCb>`

File: src/connstat.cpp [hunk 16]
```diff
@@ -0,0 +1,169 @@
...
+    /*
+     * Start receiving messages. The function nl_recvmsgs_default() will block
+     * until one or more netlink messages (notification) are received which
+     * will be passed on to nlcb().
+     * 0 on success or a negative error code from nl_recv():
+     * Number of bytes read,
+     * 0 on EOF,
+     * 0 on no data event (non-blocking mode), or
+     * a negative error code.
+     */
```
Comment (aberaud) [inline src/connstat.cpp]: remove comment

File: src/connstat.cpp [hunk 17]
```diff
@@ -0,0 +1,169 @@
...
+
+ConnectivityStatus::ConnectivityStatus()
+    : nlsk(nlsk_init())
+    , logger_(new dht::Logger)
+{
+    if (not nlsk_setup(nlsk.get()))
+        throw std::runtime_error("nlsk_setup() failed\n");
+
+    thrd_ = std::thread([this] () { nl_event_loop_thrd(nlsk.get()); });
+    thrd_.detach();
```
Comment (aberaud) [inline src/connstat.cpp]: never detach thread, especially since you capture this (would cause use-after-free).

File: src/connstat.cpp [hunk 18]
```diff
@@ -0,0 +1,169 @@
...
+            return get_neigh_state(msg);
+            break;
+        default:
+            break;
+    }
+
+    return SUCCESS;
+}
+
+static int nlcb(struct nl_msg* msg, void* arg)
```
Comment (aberaud) [inline src/connstat.cpp]: no need to split declaration/definition (could even be inlined as a lambda in call to  nl_socket_modify_cb)

File: src/connstat.cpp [hunk 19]
```diff
@@ -0,0 +1,169 @@
...
+        break;
+    default:
+        break;
+    }
+}
+
+void
+ConnectivityStatus::executer(topic topic)
+{
+    long unsigned int t = static_cast<unsigned int>(topic);
```
Comment (aberaud) [inline src/connstat.cpp]: cast from/to integer not needed

File: include/opendht/connstat.h (hunk 5 above)
Comment (257) [reply include/opendht/connstat.h]: moved to ```std::map```

File: src/connstat.cpp (hunk 17 above)
Comment (257) [reply src/connstat.cpp]: without it i get:
_terminate called without an active exception_

File: src/connstat.cpp [hunk 20]
```diff
@@ -0,0 +1,151 @@
...
+            executer(Event::ADDR);
+            executer(Event::DELADDR);
+            break;
+        case RTM_NEWNEIGH:
+        case RTM_DELNEIGH:
+        case RTM_NEWNEIGHTBL:
+            get_neigh_state(msg);
+            break;
+        default:
+            status = 1;
```
[the remark this answers is not in the document]
Comment (257) [reply src/connstat.cpp]: switched to using ```nl_cb_action```

File: include/opendht/def.h [hunk 21]
```diff
@@ -42,3 +42,5 @@
 
 // bytes
 #define HASH_LEN 20u
+
+#define OPENDHT_UNUSED_ARG(arg)  (void)arg
```
Comment (aberaud) [inline include/opendht/def.h]: remove

File: src/connstat.cpp [hunk 22]
```diff
@@ -0,0 +1,175 @@
...
+
+    thrd_ = std::thread([this] () { nl_event_loop_thrd(nlsk.get()); });
+}
+
+ConnectivityStatus::~ConnectivityStatus()
+{
+    stop = true;
+
+    nl_socket_modify_cb(nlsk.get(), NL_CB_VALID, NL_CB_CUSTOM, [](nl_msg* msg, void* data) -> int {
+            OPENDHT_UNUSED_ARG(msg);
```
Comment (aberaud) [inline src/connstat.cpp]: simply use `[](nl_msg* /*msg*/, void* data)` in argument list

File: src/connstat.cpp [hunk 23]
```diff
@@ -0,0 +1,175 @@
...
+            if (((ConnectivityStatus*)data)->logger_)
+                ((ConnectivityStatus*)data)->logger_->w("stopping msg processing\n");
```
Comment (aberaud) [inline src/connstat.cpp]: remove log

File: src/connstat.cpp [hunk 24]
```diff
@@ -0,0 +1,175 @@
...
+            return (((ConnectivityStatus*)data)->stop) ? NL_OK : NL_STOP;
+    }, (void*)this);
+
+    nl_socket_modify_cb(nlsk.get(), NL_CB_MSG_IN, NL_CB_CUSTOM, [](nl_msg* msg, void* data) -> int {
+            OPENDHT_UNUSED_ARG(msg);
+            if (((ConnectivityStatus*)data)->logger_)
+                ((ConnectivityStatus*)data)->logger_->w("got incoming msg\n");
```
Comment (aberaud) [inline src/connstat.cpp]: remove

File: src/connstat.cpp [hunk 25]
```diff
@@ -0,0 +1,175 @@
...
+        return (((ConnectivityStatus*)data)->stop) ? NL_OK : NL_STOP;
+    }, (void*)this);
+
+    nl_socket_modify_cb(nlsk.get(), NL_CB_ACK, NL_CB_CUSTOM, [](nl_msg* msg, void* data) -> int {
+            OPENDHT_UNUSED_ARG(msg);
+            if (((ConnectivityStatus*)data)->logger_)
+                ((ConnectivityStatus*)data)->logger_->w("taking care of ACK\n");
```
Comment (aberaud) [inline src/connstat.cpp]: remove

File: src/connstat.cpp [hunk 26]
```diff
@@ -0,0 +1,175 @@
...
+    case RTM_NEWNEIGH:
+    case RTM_DELNEIGH:
+    case RTM_NEWNEIGHTBL:
+        get_neigh_state(msg);
+        break;
+    default:
+        status = NL_SKIP;
+        break;
+    }
+    return status;
```
Comment (aberaud) [inline src/connstat.cpp]: simply return NL_STOP if stopping here ?

File: src/connstat.cpp [hunk 27]
```diff
@@ -0,0 +1,175 @@
...
+
+}
+
+void
+ConnectivityStatus::nl_event_loop_thrd(nl_sock *nlsk)
+{
+    int status;
```
Comment (aberaud) [inline src/connstat.cpp]: status not used

File: src/connstat.cpp (hunk 26 above)
Comment (257) [reply src/connstat.cpp]: ```NL_STOP``` actually reports error; however for ```NL_CB_VALID``` it wouldn't make a difference.

i moved us to ```NL_CB_MSG_IN``` which means more control; e.g. terminating early since we get to stop early and possibly exit before libnl starts allocate things for it's routines; ack reply for instance.
anyway, this means that we should return ```NL_OK``` so that nllib continues with its routine in case we're not terminating; continuing with next msg for multipart ones or sending back ack to kernel and so on.

a side note, the changes i'm about to push to this are not testing on termination because of the problem we saw the other day.
